src/crypto/tls/ech.go

```diff
@@ -105,327 +104,287 @@ func (c *Conn) echOfferOrGrease(helloBase *clientHelloMsg) (hello, helloInner *c
+	encodedHelloInner := echEncodeClientHelloInner(helloInner.marshal(), outerExtensions)
+	if encodedHelloInner == nil {
+		return nil, nil, errors.New("tls: ech: encoding of EncodedClientHelloInner failed")
 	}
 
-	// Compute the dummy "payload".
-	dummyEncodedHelloInnerLen := 100 // TODO(cjpatton): Compute this correctly.
-	dummyEch.payload = make([]byte, dummyEncodedHelloInnerLen+dummyAeadOverheadLen)
-	if _, err = io.ReadFull(config.rand(), dummyEch.payload); err != nil {
-		return nil, nil, fmt.Errorf("tls: ech covertext: %s", err)
+	// ClientHelloOuter is constructed by generating a fresh ClientHello and
+	// copying "key_share", "random", and "sesion_id" from helloBase, setting
+	// "server_name" to be the client-facing server, and adding the
+	// "encrypted_client_hello" extension.
+	hello, _, err = c.makeClientHello(config.MinVersion)
+	if err != nil {
+		return nil, nil, fmt.Errorf("tls: ech: %s", err)
+	}
+	hello.keyShares = helloBase.keyShares
+	hello.random = helloBase.random
+	hello.sessionId = helloBase.sessionId
+	hello.serverName = hostnameInSNI(c.ech.publicName)
+	hello.echIsOuter = true
+
+	// ClientECH, the payload of the "encrypted_client_hello" extension.
+	var ech echClient
+	_, kdfId, aeadId := c.ech.sealer.Suite().Params()
+	ech.handle.suite.kdfId = uint16(kdfId)
+	ech.handle.suite.aeadId = uint16(aeadId)
+	ech.handle.configId = configId // Empty after HRR
+	ech.handle.enc = enc           // Empty after HRR
+	helloOuterAad := echEncodeClientHelloOuterAAD(hello.marshal(), ech.handle.marshal())
+	if helloOuterAad == nil {
+		return nil, nil, errors.New("tls: ech: encoding of ClientHelloOuterAAD failed")
+	}
+	ech.payload, err = c.ech.sealer.Seal(encodedHelloInner, helloOuterAad)
+	if err != nil {
+		return nil, nil, fmt.Errorf("tls: ech: seal failed: %s", err)
 	}
+	if testingECHTriggerPayloadDecryptError {
+		ech.payload[0] ^= 0xff // Inauthentic ciphertext
+	}
+	hello.ech = ech.marshal()
 
-	// Add the dummy ECH extension to the ClientHello.
-	hello.encryptedClientHelloOffered = true
-	hello.encryptedClientHello = dummyEch.marshal()
-
-	// Bypass ECH and provide covertext.
-	c.ech.offered = false
-	c.ech.greased = true
+	// Offer ECH.
+	c.ech.offered = true
+	helloInner.raw = nil
 	hello.raw = nil
-	return hello, nil, nil
+	return hello, helloInner, nil
 }
 
-func (c *Conn) echAcceptOrBypass(hello *clientHelloMsg) (*clientHelloMsg, error) {
+func (c *Conn) echAcceptOrReject(hello *clientHelloMsg) (*clientHelloMsg, error) {
 	config := c.config
-	echProvider := config.ServerECHProvider
-
-	// Decide whether to bypass ECH.
-	echSentBeforeHrr := c.ech.offered || c.ech.greased
-	if !config.echCanAccept() ||
-		!hello.encryptedClientHelloOffered ||
-		len(hello.encryptedClientHello) == 0 {
-		if c.hrrTriggered && echSentBeforeHrr {
-			// Detect illegal second ClientHello.
+	p := config.ServerECHProvider
+
+	if !config.echCanAccept() {
+		// Bypass ECH.
+		return hello, nil
+	}
+
+	if hello.echIsInner {
+		if hello.echIsOuter {
```

Comment:
Do we have a check somewhere for the opposite case? Namely, if this is CHOuter and "ech_is_inner" is present?

Reply:
Yes, see the new test case in TestECHHandshake.

Reply:
👍